File: kattis-probs/supermario169.py
#WORK IN PROGRESS
#https://open.kattis.com/problems/supermario169
#TSP

import math
import sys

INF = float("inf")

def dist_sq(a, b):
    return (a[0]-b[0])**2 + (a[1]-b[1])**2 + (a[2]-b[2])**2

def dist(a, b):
    return math.sqrt(dist_sq(a, b))

def solve_cluster_tsp(switch_pos, coins):
    k = len(coins)
    if k == 0:
        return 0.0, [0.0], [switch_pos]
    d0 = [dist(switch_pos, c) for c in coins]
    
    # coin to coin distances
    d = [[0.0] * k for _ in range(k)]
    for i in range(k):
        for j in range(i+1, k):
            d_ij = dist(coins[i], coins[j])
            d[i][j] = d_ij
            d[j][i] = d_ij
    
    dp = [[INF] * k for _ in range(1<<k)]
    
    for i in range(k):
        dp[1<<i][i] = d0[i]
    
    for mask in range(1, 1<<k):
        # Find all possible last coins
        for last in range(k):
            if not (mask & (1<<last)) or dp[mask][last] == INF:
                continue
            
            # Try to add a new coin
            for nxt in range(k):
                if mask & (1<<nxt):
                    continue
                new_mask = mask | (1<<nxt)
                new_cost = dp[mask][last] + d[last][nxt]
                if dp[new_mask][nxt] > new_cost:
                    dp[new_mask][nxt] = new_cost
    
    full_mask = (1<<k) - 1
    
    return_to_switch = [dist(coins[i], switch_pos) for i in range(k)]
    
    # Cost to collect all coins and return to switch
    min_return_cost = INF
    for i in range(k):
        total = dp[full_mask][i] + return_to_switch[i]
        if total < min_return_cost:
            min_return_cost = total
    
    end_costs = [dp[full_mask][i] for i in range(k)]
    
    return min_return_cost, end_costs, coins

def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    
    it = iter(data)
    n = int(next(it))
    start = (int(next(it)), int(next(it)), int(next(it)))
    
    switches = []
    coin_clusters = []
    cluster_return_costs = []  # cost to collect all coins and return to switch
    cluster_end_costs = []     # for each ending coin, cost to collect all coins
    
    for _ in range(n):
        k = int(next(it))
        sx, sy, sz = int(next(it)), int(next(it)), int(next(it))
        switches.append((sx, sy, sz))
        
        coins = []
        for _ in range(k):
            cx, cy, cz = int(next(it)), int(next(it)), int(next(it))
            coins.append((cx, cy, cz))
        coin_clusters.append(coins)
        
        # Precompute for this cluster
        return_cost, end_costs, _ = solve_cluster_tsp(switches[-1], coins)
        cluster_return_costs.append(return_cost)
        cluster_end_costs.append(end_costs)
    
    # Precompute all distances between switches
    switch_dist = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for j in range(i+1, n):
            d = dist(switches[i], switches[j])
            switch_dist[i][j] = d
            switch_dist[j][i] = d
    
    # Precompute start to switch distances
    start_to_switch = [dist(start, s) for s in switches]
    
    coin_to_switch = []
    for i in range(n):
        coins = coin_clusters[i]
        k = len(coins)
        cts = [[0.0] * n for _ in range(k)]
        for coin_idx in range(k):
            for j in range(n):
                if i == j:
                    cts[coin_idx][j] = 0.0
                else:
                    cts[coin_idx][j] = dist(coins[coin_idx], switches[j])
        coin_to_switch.append(cts)

    dp = [[[INF] * (max(len(c) for c in coin_clusters) + 1) for _ in range(n)] for _ in range(1<<n)]
    
    for i in range(n):
        # Start -> switch i -> collect coins -> end at switch
        cost = start_to_switch[i] + cluster_return_costs[i]
        dp[1<<i][i][0] = min(dp[1<<i][i][0], cost)  # last_coin_idx = -1 -> index 0
        
        # Start -> switch i -> collect coins -> end at coin j
        k = len(coin_clusters[i])
        for j in range(k):
            cost = start_to_switch[i] + cluster_end_costs[i][j]
            dp[1<<i][i][j+1] = min(dp[1<<i][i][j+1], cost)  # last_coin_idx = j -> index j+1
    
    # DP transitions
    for mask in range(1, 1<<n):
        for last_switch in range(n):
            if not (mask & (1<<last_switch)):
                continue
            
            k_last = len(coin_clusters[last_switch])
            for last_pos in range(k_last + 1):  # 0=switch, 1..k_last=coins
                current_cost = dp[mask][last_switch][last_pos]
                if current_cost == INF:
                    continue
                
                if last_pos == 0:
                    current_pos = switches[last_switch]
                else:
                    current_pos = coin_clusters[last_switch][last_pos-1]
                
                # Try to go to an unvisited switch
                for next_switch in range(n):
                    if mask & (1<<next_switch):
                        continue
                    
                    # Distance from current position to next switch
                    if last_pos == 0:
                        travel = switch_dist[last_switch][next_switch]
                    else:
                        travel = dist(current_pos, switches[next_switch])
                    
                    new_mask = mask | (1<<next_switch)
                    
                    # Option 1: At next switch, collect coins and end at switch
                    cost1 = current_cost + travel + cluster_return_costs[next_switch]
                    dp[new_mask][next_switch][0] = min(dp[new_mask][next_switch][0], cost1)
                    
                    # Option 2: At next switch, collect coins and end at coin j
                    k_next = len(coin_clusters[next_switch])
                    for j in range(k_next):
                        cost2 = current_cost + travel + cluster_end_costs[next_switch][j]
                        dp[new_mask][next_switch][j+1] = min(dp[new_mask][next_switch][j+1], cost2)
    
    # Find minimum cost to visit all switches
    full_mask = (1<<n) - 1
    answer = INF
    
    for last_switch in range(n):
        for last_pos in range(len(coin_clusters[last_switch]) + 1):
            answer = min(answer, dp[full_mask][last_switch][last_pos])
    
    if n == 0:
        print("0.0000000000")
    else:
        print(f"{answer:.10f}")

if __name__ == "__main__":
    main()

File: kattis-probs/supermario169.cpp
//https://open.kattis.com/problems/supermario169

#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <limits>
#include <iomanip>
#include <sstream>

using namespace std;

const double INF = numeric_limits<double>::max();

struct Point {
    int x, y, z;
    Point() : x(0), y(0), z(0) {}
    Point(int x, int y, int z) : x(x), y(y), z(z) {}
};

double dist_sq(const Point& a, const Point& b) {
    double dx = a.x - b.x;
    double dy = a.y - b.y;
    double dz = a.z - b.z;
    return dx*dx + dy*dy + dz*dz;
}

double dist(const Point& a, const Point& b) {
    return sqrt(dist_sq(a, b));
}

struct ClusterResult {
    double return_cost;
    vector<double> end_costs;
    vector<Point> coins;
};

ClusterResult solve_cluster_tsp(const Point& switch_pos, const vector<Point>& coins) {
    int k = coins.size();
    ClusterResult result;
    result.coins = coins;
    
    if (k == 0) {
        result.return_cost = 0.0;
        result.end_costs.push_back(0.0);
        return result;
    }
    
    // Distance from switch to each coin
    vector<double> d0(k);
    for (int i = 0; i < k; i++) {
        d0[i] = dist(switch_pos, coins[i]);
    }
    
    // Distance between coins
    vector<vector<double>> d(k, vector<double>(k, 0.0));
    for (int i = 0; i < k; i++) {
        for (int j = i + 1; j < k; j++) {
            double d_ij = dist(coins[i], coins[j]);
            d[i][j] = d_ij;
            d[j][i] = d_ij;
        }
    }
    
    // DP for TSP within cluster
    int full_mask_size = 1 << k;
    vector<vector<double>> dp(full_mask_size, vector<double>(k, INF));
    
    // Initialize with starting at each coin
    for (int i = 0; i < k; i++) {
        dp[1 << i][i] = d0[i];
    }
    
    // DP transitions
    for (int mask = 1; mask < full_mask_size; mask++) {
        for (int last = 0; last < k; last++) {
            if (!(mask & (1 << last)) || dp[mask][last] == INF) {
                continue;
            }
            
            for (int nxt = 0; nxt < k; nxt++) {
                if (mask & (1 << nxt)) {
                    continue;
                }
                int new_mask = mask | (1 << nxt);
                double new_cost = dp[mask][last] + d[last][nxt];
                if (dp[new_mask][nxt] > new_cost) {
                    dp[new_mask][nxt] = new_cost;
                }
            }
        }
    }
    
    int full_mask = (1 << k) - 1;
    
    // Cost to return to switch from each coin
    vector<double> return_to_switch(k);
    for (int i = 0; i < k; i++) {
        return_to_switch[i] = dist(coins[i], switch_pos);
    }
    
    // Minimum cost to collect all coins and return to switch
    double min_return_cost = INF;
    for (int i = 0; i < k; i++) {
        double total = dp[full_mask][i] + return_to_switch[i];
        if (total < min_return_cost) {
            min_return_cost = total;
        }
    }
    result.return_cost = min_return_cost;
    
    // Cost to end at each coin (without returning to switch)
    result.end_costs.resize(k);
    for (int i = 0; i < k; i++) {
        result.end_costs[i] = dp[full_mask][i];
    }
    
    return result;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n;
    cin >> n;
    
    Point start;
    cin >> start.x >> start.y >> start.z;
    
    vector<Point> switches(n);
    vector<vector<Point>> coin_clusters(n);
    vector<double> cluster_return_costs(n);
    vector<vector<double>> cluster_end_costs(n);
    
    // Read input and precompute cluster solutions
    for (int i = 0; i < n; i++) {
        int k;
        cin >> k;
        cin >> switches[i].x >> switches[i].y >> switches[i].z;
        
        vector<Point> coins(k);
        for (int j = 0; j < k; j++) {
            cin >> coins[j].x >> coins[j].y >> coins[j].z;
        }
        coin_clusters[i] = coins;
        
        // Solve TSP for this cluster
        ClusterResult cluster_result = solve_cluster_tsp(switches[i], coins);
        cluster_return_costs[i] = cluster_result.return_cost;
        cluster_end_costs[i] = cluster_result.end_costs;
    }
    
    if (n == 0) {
        cout << "0.0000000000\n";
        return 0;
    }
    
    // Precompute distances between switches
    vector<vector<double>> switch_dist(n, vector<double>(n, 0.0));
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            double d = dist(switches[i], switches[j]);
            switch_dist[i][j] = d;
            switch_dist[j][i] = d;
        }
    }
    
    // Precompute distances from start to each switch
    vector<double> start_to_switch(n);
    for (int i = 0; i < n; i++) {
        start_to_switch[i] = dist(start, switches[i]);
    }
    
    // Precompute distances from coins to switches
    vector<vector<vector<double>>> coin_to_switch(n);
    for (int i = 0; i < n; i++) {
        int k = coin_clusters[i].size();
        vector<vector<double>> cts(k, vector<double>(n, 0.0));
        for (int coin_idx = 0; coin_idx < k; coin_idx++) {
            for (int j = 0; j < n; j++) {
                if (i == j) {
                    cts[coin_idx][j] = 0.0;
                } else {
                    cts[coin_idx][j] = dist(coin_clusters[i][coin_idx], switches[j]);
                }
            }
        }
        coin_to_switch[i] = cts;
    }
    
    // Find maximum number of coins in any cluster
    int max_coins = 0;
    for (const auto& cluster : coin_clusters) {
        max_coins = max(max_coins, (int)cluster.size());
    }
    
    // DP: dp[mask][last_switch][last_pos]
    // last_pos = 0: at switch, last_pos > 0: at coin (last_pos-1)
    int dp_size = 1 << n;
    vector<vector<vector<double>>> dp(
        dp_size, 
        vector<vector<double>>(
            n, 
            vector<double>(max_coins + 1, INF)
        )
    );
    
    // Initialize DP
    for (int i = 0; i < n; i++) {
        // Start -> switch i -> collect coins -> end at switch
        double cost = start_to_switch[i] + cluster_return_costs[i];
        dp[1 << i][i][0] = min(dp[1 << i][i][0], cost);
        
        // Start -> switch i -> collect coins -> end at coin j
        int k = coin_clusters[i].size();
        for (int j = 0; j < k; j++) {
            double cost = start_to_switch[i] + cluster_end_costs[i][j];
            dp[1 << i][i][j + 1] = min(dp[1 << i][i][j + 1], cost);
        }
    }
    
    // DP transitions
    for (int mask = 1; mask < dp_size; mask++) {
        for (int last_switch = 0; last_switch < n; last_switch++) {
            if (!(mask & (1 << last_switch))) {
                continue;
            }
            
            int k_last = coin_clusters[last_switch].size();
            for (int last_pos = 0; last_pos <= k_last; last_pos++) {
                double current_cost = dp[mask][last_switch][last_pos];
                if (current_cost == INF) {
                    continue;
                }
                
                Point current_pos;
                if (last_pos == 0) {
                    current_pos = switches[last_switch];
                } else {
                    current_pos = coin_clusters[last_switch][last_pos - 1];
                }
                
                // Try to go to an unvisited switch
                for (int next_switch = 0; next_switch < n; next_switch++) {
                    if (mask & (1 << next_switch)) {
                        continue;
                    }
                    
                    // Distance from current position to next switch
                    double travel;
                    if (last_pos == 0) {
                        travel = switch_dist[last_switch][next_switch];
                    } else {
                        travel = dist(current_pos, switches[next_switch]);
                    }
                    
                    int new_mask = mask | (1 << next_switch);
                    
                    // Option 1: At next switch, collect coins and end at switch
                    double cost1 = current_cost + travel + cluster_return_costs[next_switch];
                    if (dp[new_mask][next_switch][0] > cost1) {
                        dp[new_mask][next_switch][0] = cost1;
                    }
                    
                    // Option 2: At next switch, collect coins and end at coin j
                    int k_next = coin_clusters[next_switch].size();
                    for (int j = 0; j < k_next; j++) {
                        double cost2 = current_cost + travel + cluster_end_costs[next_switch][j];
                        if (dp[new_mask][next_switch][j + 1] > cost2) {
                            dp[new_mask][next_switch][j + 1] = cost2;
                        }
                    }
                }
            }
        }
    }
    
    // Find minimum cost to visit all switches
    int full_mask = (1 << n) - 1;
    double answer = INF;
    
    for (int last_switch = 0; last_switch < n; last_switch++) {
        int k_last = coin_clusters[last_switch].size();
        for (int last_pos = 0; last_pos <= k_last; last_pos++) {
            answer = min(answer, dp[full_mask][last_switch][last_pos]);
        }
    }
    
    cout << fixed << setprecision(10) << answer << "\n";
    
    return 0;
}